### PR TITLE
Add Cluster field in ObjectMeta

### PR DESCRIPTION
**Pull request by jianhuiz**
There will be no sub-rs, but add `Cluster` field to the ObjectMeta (for all the objects)

"To distinguish the object at the federation level from it's constituents at the cluster level we will add a "Cluster" field to the metadata of all objects (where the federation itself will also have a cluster identifier).  That way it is possible to list, interact with, and distinguish between the objects either at the federation level or at the individual cluster level based on the cluster identifier. "

@quinton-hoole @nikhiljindal @deepak-vij @mfanjie @huangyuqi

[Analytics]()

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28921)

### CODE REVIEW

**Comment by kargakis** [comment]
@janetkuo since your changes we don't care about objectmeta changes anymore for deployments, right?

File: pkg/api/types.go [hunk 1]
```diff
@@ -150,6 +150,11 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty"`
+
+	// The cluster which the object belongs to. The federation apiserver/controllers use this to
+	// distinguish the object at the federation level from it's constituents at the cluster level.
+	// An empty value means the object belongs to the local cluster (the one the client connects to).
```
Comment (kshafiee) [inline pkg/api/types.go]: Is the federation-controller always supposed to be in the form of 2 pods, federation-apiserver pod and federation-controller-manager pod, in one of the federated clusters? If it is deployed on a separate VM, e.g., controller VM, it means the controller does not belong to any cluster. So, no local cluster exists and an RS with an empty cluster id value cannot be created.

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

File: pkg/api/types.go [hunk 2]
```diff
@@ -150,6 +150,11 @@ type ObjectMeta struct {
...
+	Cluster string `json:"cluster,omitempty"`
```
Comment (smarterclayton) [inline pkg/api/types.go]: If this refers to a resource by name, typically we would call this "ClusterName"

File: pkg/api/types.go (hunk 1 above)
Comment (jianhuiz) [inline pkg/api/types.go]: by `local cluster` I mean the cluster that a client connects to, not the one the apiserver (or other component) running in.

File: pkg/api/types.go (hunk 2 above)
Comment (jianhuiz) [inline pkg/api/types.go]: So change it to `ClusterName`? @quinton-hoole @nikhiljindal

**Comment by quinton-hoole** [comment]
This should ideally be an object reference of some sort (to the registered cluster), rather than just a string name. @nikhiljindal do we have such a thing available?  Can I leave the remainder of this review to you?

**Comment by smarterclayton** [comment]
Generally we don't actually do object references to cluster scoped
resources.  Arguably this isn't cluster scoped, but not much different than
"nodeName".

File: pkg/api/types.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/api/types.go]: Yes ClusterName sounds good

**Comment by nikhiljindal** [comment]
@jianhuiz We also need logic to set this field.
federation-apiserver should set this to FEDERATION_NAME before storing the objects in federation etcd.
And federation controllers (there is just service controller for now) should set this field in the objects it creates in underlying k8s clusters.

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -187,6 +187,11 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty" patchStrategy:"merge" protobuf:"bytes,14,rep,name=finalizers"`
+
+	// The cluster which the object belongs to. The federation apiserver/controllers use this to
+	// distinguish the object at the federation level from it's constituents at the cluster level.
+	// An empty value means the object belongs to the local cluster (the one the client connects to).
+	Cluster string `json:"cluster,omitempty" protobuf:"bytes,15,opt,name=cluster"`
```
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Add "Cannot be updated" as we say for Name and Namespace.
Will also need to add logic to ensure that it is not updated.

**Comment by jianhuiz** [comment]
changed to `Clustername`; added validation to prevent updating

**Comment by jianhuiz** [comment]
@nikhiljindal I'll deal with them in a separate PR.
Regarding storing the `ClusterName` in etcd, I'm still wondering whether or not it is needed to be stored in etcd as the client can actually fill this field when it gets objects from the clusters

> 
We also need logic to set this field.
federation-apiserver should set this to FEDERATION_NAME before storing the objects in federation etcd.
And federation controllers (there is just service controller for now) should set this field in the objects it creates in underlying k8s clusters.
>
Its fine if you plan to add that in a separate PR.

**Comment by smarterclayton** [comment]
IF we don't want to store clusterName, we'll want to use the generic
registry Decorator pattern for that.

**Comment by quinton-hoole** [comment]
@smarterclayton I think we can consider this to be a cluster-scoped resource.  I'm all for consistency, but there a good reason why "Generally we don't actually do object references to cluster scoped resources."?   I personally hate loosely typed strings floating around, as they have been the source of a whole lot of nasty bugs.  Other than consistency with the past, is there any argument for preferring a loose string name over a more well defined object reference?

**Comment by smarterclayton** [comment]
In general when the scope is unambiguous we don't do something like an
object reference (for instance, nodeName).  We do use those extra fields in
object reference to provide qualifying info where necessary, but we aren't
slavish about having all the fields specified.  In the case of cluster, by
definition every reference is out of the current scope, and so I can buy
that we may in the future need additional qualifiers to locate the right
cluster.

I'm not opposed to `Cluster ClusterReference` with a single field `Name`.

**Comment by quinton-hoole** [comment]
@smarterclayton Agreed, that's a good solution.

**Comment by nikhiljindal** [comment]
Yes adding `Cluster ClusterReference` with a single field `Name` sounds good.

Re: whether we should store it in etcd or not: I discussed this with @lavalamp and he suggested that its fine to store it in etcd, but if the cluster identifier can change and the same cluster can have different names as part of different federations, then it might be better to NOT store it in etcd and federation apiserver can populate that field before returning the resource.
@jianhuiz Had also mentioned a similar concern.

Also the kubernetes cluster itself does not have a global unique immutable identifier that it knows about. If we had such a thing, then the kubernetes apiserver could set that in the Cluster field at resource creation time. We could store that in etcd.

This is what we can do now:
* In federation and kubernetes apiservers: Ignore and clear this field if it is set when the resource is being created/updated.
* In federation-apiserver: when it reads resources from underlying clusters, set this field to refer the cluster from which that resource was read.
* In federation-apiserver: when it reads resources from its own etcd, set this field as Name = FEDERATION_NAME.

The aim of this field is to distinguish resources when users get resources from all kubernetes clusters via the federation apiserver. There will be multiple resources with the same name and namespace. We want to add the cluster field, so that resources have unique combination of name, namespace and cluster.

cc @bgrant0607

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -187,6 +187,12 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty" patchStrategy:"merge" protobuf:"bytes,14,rep,name=finalizers"`
+
+	// Name of the cluster which the object belongs to. The federation apiserver/controllers use this to
```
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: I think we can do without mentioning "federation" here.
Just say that this is used to distinguish resources with same name and namespace in different clusters.

**Comment by jianhuiz** [comment]
Yes, if we were to add other fields into ClusterReference, the re-population of this is more likely needed to happen. Besides, it is always need to populate it to the existing objects when the k8s joins into a federation for the first time. Save it into etcd also requires upgrade an existing k8s cluster before joining it into a federation

**Comment by mwielgus** [comment]
@jianhuiz - When do you think you will be able to address the comments? This field will be quite helpful in #29347 - otherwise I will have to do some temporary workarounds.

**Comment by jianhuiz** [comment]
@mwielgus working on it, will push in a couple hours.

**Comment by nikhiljindal** [comment]
@mwielgus curious how are you planning to use this Cluster field in the library?

**Comment by mfanjie** [comment]
One further comment on this design, for federation level object like federation service, it contains some aggregated properties like LB IP from different underlying clusters, even with this cluster field in object meta, how can we tell which LB comes from which cluster? how this issue #26329 can be resolved?

**Comment by mwielgus** [comment]
@nikhiljindal - not in the library itself but in a sample/reference controller (namespace). The library functions makes an assumption that they can return an aggregated list of objects from all clusters and it will be possible to tell which one came from which. Obviously without cluster name in objectmeta some other temporary way should be established (like hacky annotations) to substitute it.

**Comment by mwielgus** [comment]
@jianhuiz ping

**Comment by jianhuiz** [comment]
rebased

**Comment by jianhuiz** [comment]
rebased

**Comment by nikhiljindal** [comment]
@jianhuiz Want to make sure that you are working on the changes we discussed and not blocked on me.

I see that you have updated the API changes (to use ClusterReference), but still need to make sure that it is not stored in etcd.
Please let me know when you are done with all the changes. Thx.

File: api/swagger-spec/apps_v1alpha1.json [hunk 5]
```diff
@@ -1145,6 +1149,16 @@
      }
     }
    },
+   "v1.ClusterReference": {
+    "id": "v1.ClusterReference",
+    "description": "ClusterReference contains information to identify a cluster",
+    "properties": {
+     "clusterName": {
+      "type": "string",
+      "description": "The name of the cluster Cannot be updated."
```
Comment (nikhiljindal) [inline api/swagger-spec/apps_v1alpha1.json]: Add a "." before Cannot

File: pkg/api/types.go [hunk 6]
```diff
@@ -150,6 +150,12 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty"`
+
+	// A referece to the cluster which the object belongs to.
+	// This is used to distinguish resources with same name and namespace in different clusters.
+	// An empty value means the object belongs to the local cluster (the one the client connects to).
+	// Cannot be updated.
+	Cluster *ClusterReference `json:"cluser,omitempty"`
```
Comment (nikhiljindal) [inline pkg/api/types.go]: json field name is incorrect.
cluser -> cluster

File: pkg/api/validation/validation.go [hunk 7]
```diff
@@ -242,6 +252,11 @@ var ValidateServiceAccountName = NameIsDNSSubdomain
 // trailing dashes are allowed.
 var ValidateEndpointsName = NameIsDNSSubdomain
 
+// ValidateClusterName can be used to check whether the given cluster name is valid.
+// Prefix indicates this name will be used as part of generation, in which case
```
Comment (nikhiljindal) [inline pkg/api/validation/validation.go]: We never use it to generate name. Delete this line.

File: federation/apis/federation/validation/validation.go [hunk 8]
```diff
@@ -22,8 +22,6 @@ import (
 	"k8s.io/kubernetes/pkg/util/validation/field"
 )
 
-var ValidateClusterName = validation.NameIsDNSSubdomain
```
Comment (thockin) [inline federation/apis/federation/validation/validation.go]: Can we change Cluster name to a DNS label?  Does it need to be `foo.bar.bat` or is `foo-bar-bat` good enough?  Making it a subdomain makes it very hard to use this name in DNS ever,

**Comment by smarterclayton** [comment]
I thought we defined cluster names as subdomains a while back.  I could
imagine clusters having multiple segments for lots of reasons, since we
don't have a grouping concept for clusters.

**Comment by jianhuiz** [comment]
fixed and rebased

**Comment by thockin** [comment]
If cluster names are subdomains, just know that we can never practically
use them in a DNS name.  We might get away with it sometimes but there are
too many limits that add up to overflow and search paths are already too
messy.

**Comment by janetkuo** [comment]
> @janetkuo since your changes we don't care about objectmeta changes anymore for deployments, right?

@kargakis if you meant pod-template-hash, then yes!

**Comment by nikhiljindal** [comment]
It will be good to be able to use cluster names in DNS names. We can use labels for grouping.

I am inclined towards defining cluster names as domain names instead of subdomain names.
@quinton-hoole What do you think?

**Comment by nikhiljindal** [comment]
from @smarterclayton:
> I thought we defined cluster names as subdomains a while back.

Actually thats true. We allow subdomains as cluster names in federation/v1beta1/cluster: https://github.com/kubernetes/kubernetes/blob/d412d5721d4d2c5b92d142eb02c9203187bb3f9b/federation/apis/federation/validation/validation.go#L25.

**Comment by thockin** [comment]
Is it too late to change.  DNS "Labels" are preferred, IMO, and It was a
mistake to allow anything to be named as DNS Subdomains without explicit
justification

**Comment by nikhiljindal** [comment]
@jianhuiz I dont see any code to prevent users from setting the ClusterReference field while creating a resource. Did I miss something?
Please also add a test verifying that a create request fails if it has ClusterReference set.

from @smarterclayton:

> IF we don't want to store clusterName, we'll want to use the generic
> registry Decorator pattern for that.

Can you point to an example @smarterclayton? Another way would be to return an error during validation if a create request contains ClusterReference. I see thats what we do for root scoped resources containing namespaces: https://github.com/kubernetes/kubernetes/blob/d412d5721d4d2c5b92d142eb02c9203187bb3f9b/pkg/api/validation/validation.go#L327. We return a field forbidden error.

**Comment by smarterclayton** [comment]
You don't think representing the location of the cluster as its name
is appropriate?

**Comment by jianhuiz** [comment]
@nikhiljindal I thought we are going to make the apiserver ignore this field, however put it in strategy is better. I think we should empty this field and continue instead of fail it?
> I dont see any code to prevent users from setting the ClusterReference field while creating a resource. Did I miss something?
> Please also add a test verifying that a create request fails if it has ClusterReference set.

**Comment by nikhiljindal** [comment]
> @nikhiljindal I thought we are going to make the apiserver ignore this field, however put it in strategy is better. I think we should empty this field and continue instead of fail it?

I think returning an error is better than silently removing it.
If my create request succeeds with cluster field set, then I expect to get back an object with cluster field set to the same value on GET.

Also, I talked to @lavalamp and adding validation to confirm that the field is not set should be enough. Dont need to do anything special in the registry.

**Comment by jianhuiz** [comment]
@nikhiljindal
I see your point, but it would mean an update operation need to 
1. get the object and update interested field
__2. clear cluster field__
3. do update

File: pkg/api/types.go [hunk 9]
```diff
@@ -150,6 +150,12 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty"`
+
+	// A referece to the cluster which the object belongs to.
+	// This is used to distinguish resources with same name and namespace in different clusters.
+	// An empty value means the object belongs to the local cluster (the one the client connects to).
```
Comment (lavalamp) [inline pkg/api/types.go]: I think at some point this may stop being true, clusters will present their own IDs.

File: pkg/api/types.go [hunk 10]
```diff
@@ -165,6 +171,13 @@ const (
 	TerminationMessagePathDefault string = "/dev/termination-log"
 )
 
+// ClusterReference contains information to identify a cluster.
+type ClusterReference struct {
+	// The name of the cluster.
+	// Cannot be updated.
```
Comment (lavalamp) [inline pkg/api/types.go]: If the name is immutable, then it should be an ID.

I also think that this needs a ClusterUID field.

Comment (lavalamp) [inline pkg/api/types.go]: I think @bgrant0607 needs to approve this change, since it affects every object in the system.

**Comment by bgrant0607** [comment]
For context:

I suggested adding a cluster identifier to metadata in order to distinguish ReplicaSet Foo in namespace Bar in cluster A from ReplicaSet Foo in namespace Bar in cluster B from ReplicaSet Foo in namespace Bar in federation F.

With regards to what kind of identifier, since users will need to read and write them, a cluster name would be more user-friendly than a uid. That said, I fully expect to add a clusterUid, as well. The need for a cluster uid has arisen in other contexts (e.g., https://github.com/kubernetes/features/issues/50). 

I would like clusters to know their own names and for those names to be consistent with names in kubeconfig (or its successor).

Yes, there is no way to guarantee global uniqueness of names, but I'm comfortable pushing that responsibility onto the users building a federation.

So, the cluster name in the metadata is not a reference, no more than the namespace is a reference. It is the scope for that resource.

I'm also not keen to add additional nesting within metadata. That seems unnecessary. We've been discussing the addition of namespaceUid, and that wouldn't be nested with namespace.

@smarterclayton What did you mean by "location of the cluster"?

**Comment by nikhiljindal** [comment]
> For context:
> 
> I suggested adding a cluster identifier to metadata in order to distinguish ReplicaSet Foo in namespace Bar in cluster A from ReplicaSet Foo in namespace Bar in cluster B from ReplicaSet Foo in namespace Bar in federation F.

Yes, thats the intent of this field.

> With regards to what kind of identifier, since users will need to read and write them, a cluster name would be more user-friendly than a uid. That said, I fully expect to add a clusterUid, as well. The need for a cluster uid has arisen in other contexts (e.g., kubernetes/features#50).

Why will users have to write them? To indicate which cluster should that resource end up in?
In federation, we want a mechanism for users to specify a subset of clusters in which their resources should be created (by default they are spread across all registered clusters): https://github.com/kubernetes/kubernetes/issues/29887. The current proposal is to use LabelSelectors for that which is much more flexible than this Cluster field.
What if we keep that as the canonical way to decide which cluster should your resource end up in and say that users can never set the Cluster field. They can only read it to know which cluster is the resource coming from?

> I would like clusters to know their own names and for those names to be consistent with names in kubeconfig (or its successor).

FWIW, right now, we cant use kubeconfig names as is. Cluster name right now is supposed to be a [valid DNS1123Subdomain](https://github.com/kubernetes/kubernetes/blob/0410c339954f2e81a39d4f1dc88a7c36a7d7c98c/pkg/util/validation/validation.go#L111) as most other kube resources which does not allow underscores and the length has to be less than 253 chars. cluster names in kubeconfig dont confirm to this. We can probably try to ensure that in the kubeconfig successor.

Also, what we were planning to do right now is: add this Cluster field to the API, but kube-apiserver will not set it in GET and ensure that it is not set by the user in Create and update requests, so that it is never stored in etcd.
Federation apiserver will set this field while surfacing kubernetes resources. It will ignore this field in create and delete.
In future, when we have kube clusters which are aware of their name, we can start setting this field in kube-apiserver.
Does that sound reasonable?

> Yes, there is no way to guarantee global uniqueness of names, but I'm comfortable pushing that responsibility onto the users building a federation.
> 
> So, the cluster name in the metadata is not a reference, no more than the namespace is a reference. It is the scope for that resource.
> 
> I'm also not keen to add additional nesting within metadata. That seems unnecessary. We've been discussing the addition of namespaceUid, and that wouldn't be nested with namespace.
> 
> @smarterclayton What did you mean by "location of the cluster"?

**Comment by lavalamp** [comment]
@nikhiljindal: there's two userbases for this. Your own, in federation
context, and everyone else, ordinary kubernetes users. We need to make sure
it is natural and meaningful for both use cases.

**Comment by nikhiljindal** [comment]
Summarizing my discussion with @bgrant0607:
* No nesting: Remove ClusterReference and replace it with a string ClusterName field (which is what this PR initially had). We expect to add a ClusterUid in future.
* Both kube-apiserver and federation-apiserver will ignore it (clear the field) in create and update requests, so that the value is never stored in etcd. This is better than returning an error if it is set, since it allows compatibility between different client and server versions.
* Right now, no one will be setting this field. In 1.5, when we start surfacing kubernetes resources in federation-apiserver, then federation-apiserver will set it appropriately. And in future, when we make kubernetes clusters aware of their name, then they will start setting this as well.
* Re: using this field to indicate which cluster should the resource be created in: If the user knows which cluster they want to create their resource in, then this is clearer and more definitive than ClusterSelector. So we might allow this in future. But for now, apiserver will just ignore it if it is set in create/update request.

So the next steps for this PR to get in are:
* Update the API to replace ClusterReference by ClusterName string field.
* Add clear documentation saying: this field is not set anywhere right now. And apiserver is going to ignore it if it is set in create or update request.
* Ensure that the field is cleared in create and update requests.
* Add tests to ensure that the field is ignored and not saved in etcd.

cc @kubernetes/sig-cluster-federation @quinton-hoole @jianhuiz @lavalamp @smarterclayton @bgrant0607

**Comment by lavalamp** [comment]
That plan SGTM

**Comment by lavalamp** [comment]
btw, one further thought: cluster names as DNS subdomains makes a lot of sense to me. Presumably you guys thought of a good reason to be more permissive?

**Comment by thockin** [comment]
You mean as "labels" (a subdomain is a set of labels) - as defined in RFC :)

**Comment by nikhiljindal** [comment]
Yes I confirmed with @lavalamp, he meant labels. Have sent https://github.com/kubernetes/kubernetes/pull/30956 to update cluster name validation to allow only DNS labels.

**Comment by lavalamp** [comment]
Yeah, I apparently meant labels.

**Comment by jianhuiz** [comment]
done with new plan

**Comment by jianhuiz** [comment]
rebased

**Comment by jianhuiz** [comment]
rebased

File: pkg/api/types.go [hunk 11]
```diff
@@ -149,6 +149,12 @@ type ObjectMeta struct {
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
 	Finalizers []string `json:"finalizers,omitempty"`
+
+	// The name of the cluster which the object belongs to.
+	// This is used to distinguish resources with same name and namespace in different clusters.
+	// This field is not set anywhere right now and apiserver is going to ignore it if set in create or update request.
+	// Cannot be updated.
```
Comment (nikhiljindal) [inline pkg/api/types.go]: We can delete this "Cannot be updated" line, since we say that it is ignored in create and update requests in the previous line.

**Comment by nikhiljindal** [comment]
Thanks @jianhuiz 
Have just one comment regarding documentation.

Also, we should have an integration or e2e test that verifies that we are not storing ClusterName in etcd. We can use the etcd client as we do in other integration tests (https://github.com/kubernetes/kubernetes/blob/c0e79d8da731e52312451f017cb44240831576d4/test/integration/framework/etcd_utils.go#L43) to read etcd directly. I am with fine with you adding the integration or e2e test in another PR, seeing the number of times you have rebased this PR already.

@bgrant0607 Will be great to get an LGTM from you as well.

**Comment by jianhuiz** [comment]
fixed and rebased, will create a separate e2e PR

**Comment by bgrant0607** [comment]
types.go changes LGTM

**Comment by nikhiljindal** [comment]
Thanks @bgrant0607 
I dont think we need release note for this. There is no user visible change here.

Adding LGTM label. Thanks @jianhuiz

**Comment by nikhiljindal** [comment]
@k8s-bot e2e test this issue: #IGNORE

**Comment by jianhuiz** [comment]
rebased

**Pull request merged**